lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -354,22 +389,39 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
         -> IO (FollowAction ())
     forward blocks (_nodeTip, _pparams) = do
         atomically $ forM_ blocks $ \blk -> do
-            let (slot, registrations) = fromShelleyBlock' getEpochLength blk
-            runExceptT (putPoolProduction (getHeader blk) (getProducer blk)) >>= \case
-                Left e   -> liftIO $ traceWith tr $ MsgErrProduction e
-                Right () -> pure ()
-            forM_ registrations $ \case
-                Registration pool -> do
-                    liftIO $ traceWith tr $ MsgStakePoolRegistration pool
-                    putPoolRegistration slot pool
-                Retirement cert -> do
+            let (slot, certificates) = fromShelleyBlock' getEpochLength blk
+            runExceptT (putPoolProduction (getHeader blk) (getProducer blk))
+                >>= \case
+                    Left e ->
+                        liftIO $ traceWith tr $ MsgErrProduction e
+                    Right () ->
+                        pure ()
+
+            -- A Shelley block can contain multiple certificates relating to
+            -- the same pool.
+            --
+            -- The order in which certificates appear within a block determines
+            -- their precedence relative to one another.
+            --
+            -- Certificates that appear later in a block have higher precedence
+            -- than certificates that appear earlier.
+            --
+            let certificateIndices = [minBound ..]
+            forM_ (certificateIndices `zip` certificates) $ \case
+                (slotInternalIndex, Registration cert) -> do
+                    liftIO $ traceWith tr $ MsgStakePoolRegistration cert
+                    putPoolRegistration (slot, slotInternalIndex) cert
```

Comment:
isn't this slot internal index superfluous? 

a) All certificates within a block are inserted _atomically_, so they're either all inserted, or none of them are. 
b) A certificate present earlier in a block will be overridden by one later in the block
c) This is a strategy we've been using so far for all certificates, so I find the discrepancy here rather confusing. It suggests there's something special about these certificates, although there isn't (or am I missing something?) 

I'd suggest to drop the `slotInternalIndex` to not add on top of the existing complexity.

Reply:
> isn't this slot internal index superfluous?

It turns out that:

1. A _single block_ can contain multiple certificates relating to the same pool.
2. A _single transaction_ can contain multiple certificates relating to the same pool.
3. The _relative order_ of these certificates is _significant_.
4. In general, earlier certificates are made obsolete by later certificates, but in particular:
    - a registration certificate will always _supersede_ an earlier retirement certificate (making that earlier retirement certificate obsolete);
    - a retirement certificate will always _augment_ the latest known registration certificate.

See [this thread](https://input-output-rnd.slack.com/archives/CCRB7BU8Y/p1593584824393100) for more information.

After discussing this with @rvl and @Anviking, we identified two potential methods of dealing with this issue:
- **Option A**: when processing a given block in a given slot **_s_**, if there are multiple registration and/or retirement certificates for a given pool **_p_**, then only record the latest registration and retirement certificates for that pool in the database, and ignore all prior certificates.
- **Option B**: when processing a given block in a given slot **_s_**, always record _all_ certificates for a given pool **_p_**, but also record the _order_ in which they appeared with the block. When processing a query for a given pool, fetch the latest registration and retirement certificates, and at _that point_ determine the precedence.

Comparing these two options:
* Option A has the advantage of retaining a simpler database design.
* Option B has the advantage of retaining more information to aid with debugging.

On balance, we felt that while option B requires more invasive changes to the pool database (an extra column in three tables to record the intra-slot order), recording all the pool certificates we encounter in a block will give us more information to use in the event that we need to perform debugging.